src/ast/passes/semantic_analyser.h

```diff
@@ -159,7 +159,8 @@ class SemanticAnalyser : public Visitor {
   };
   void accept_statements(StatementList &stmts);
 
-  Scope *scope_;
+  std::vector<Node *> scope_stack_;
```

Comment:
Representing the scope stack with `std::vector` means that we're always iterating it in the "opposite" order, i.e. the outermost scope first. It shouldn't matter now as we don't support variable shadowing but if we ever decide to do so, this may become a problem so I'd at least explicitly state the fact here.

Reply:
Good call.